bundles/org.openhab.binding.sensibo/src/main/java/org/openhab/binding/sensibo/internal/handler/SensiboSkyHandler.java

```diff
@@ -0,0 +1,438 @@
+import org.apache.commons.lang.WordUtils;
+import org.apache.commons.lang.builder.ToStringBuilder;
+import org.apache.commons.lang.builder.ToStringStyle;
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
+import org.eclipse.smarthome.core.library.types.DecimalType;
+import org.eclipse.smarthome.core.library.types.OnOffType;
+import org.eclipse.smarthome.core.library.types.QuantityType;
+import org.eclipse.smarthome.core.library.types.StringType;
+import org.eclipse.smarthome.core.library.unit.SIUnits;
+import org.eclipse.smarthome.core.thing.Channel;
+import org.eclipse.smarthome.core.thing.ChannelUID;
+import org.eclipse.smarthome.core.thing.Thing;
+import org.eclipse.smarthome.core.thing.ThingStatus;
+import org.eclipse.smarthome.core.thing.binding.ThingHandlerService;
+import org.eclipse.smarthome.core.thing.binding.builder.ChannelBuilder;
+import org.eclipse.smarthome.core.thing.type.ChannelType;
+import org.eclipse.smarthome.core.thing.type.ChannelTypeBuilder;
+import org.eclipse.smarthome.core.thing.type.ChannelTypeProvider;
+import org.eclipse.smarthome.core.thing.type.ChannelTypeUID;
+import org.eclipse.smarthome.core.thing.type.StateChannelTypeBuilder;
+import org.eclipse.smarthome.core.types.Command;
+import org.eclipse.smarthome.core.types.RefreshType;
+import org.eclipse.smarthome.core.types.StateDescriptionFragmentBuilder;
+import org.eclipse.smarthome.core.types.StateOption;
+import org.eclipse.smarthome.core.types.UnDefType;
+import org.openhab.binding.sensibo.internal.CallbackChannelsTypeProvider;
+import org.openhab.binding.sensibo.internal.SensiboBindingConstants;
+import org.openhab.binding.sensibo.internal.config.SensiboSkyConfiguration;
+import org.openhab.binding.sensibo.internal.model.SensiboModel;
+import org.openhab.binding.sensibo.internal.model.SensiboSky;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import tec.uom.se.unit.Units;
+
+/**
+ * The {@link SensiboSkyHandler} is responsible for handling commands, which are
+ * sent to one of the channels.
+ *
+ * @author Arne Seime - Initial contribution
+ */
+@NonNullByDefault
+public class SensiboSkyHandler extends SensiboBaseThingHandler implements ChannelTypeProvider {
+    public static final String SWING_PROPERTY = "swing";
+    public static final String MASTER_SWITCH_PROPERTY = "on";
+    public static final String FAN_LEVEL_PROPERTY = "fanLevel";
+    public static final String MODE_PROPERTY = "mode";
+    public static final String TARGET_TEMPERATURE_PROPERTY = "targetTemperature";
+    private final Logger logger = LoggerFactory.getLogger(SensiboSkyHandler.class);
+    private Optional<SensiboSkyConfiguration> config = Optional.empty();
+    private final Map<ChannelTypeUID, ChannelType> generatedChannelTypes = new HashMap<>();
+
+    public SensiboSkyHandler(final Thing thing) {
+        super(thing);
+    }
+
+    private String getMacAddress() {
+        if (config.isPresent()) {
```

Comment:
This code is the result of the previous review. AFAIK the problem arises because the config loading mechanism does not guarantee that a config instance will be returned. Maybe the null annotations should be better aligned? 

The macAddress is a mandatory property, so my understanding based on other review comments is that initialize() won't be called by the framework unless macAddress is present?

Reply:
@seime, I think the point @martinvw makes here is that an empty mac seems to be invalid and therefore an empty mac should result in an exception rather than an empty string. 

But the main cause of the problem is the decison to forbid `@NonNullByDefault({})` for fields where the framework makes sure they are initialized when needed. This creates tons of unnecessary code.

Reply:
Changed to throwing an exception